Error: `could not find application file: erlang_v8.app`

In my personal project, after I put `{:elixir_v8, "~> 0.2.2"}` as deps in mix.exs, hit `$ mix deps.get` and `$ iex -S mix`, I see this:

```
Erlang/OTP 19 [erts-8.2] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Compiling 1 file (.ex)
Generated elixir_v8_try app

=INFO REPORT==== 16-Feb-2017::17:21:05 ===
    application: logger
    exited: stopped
    type: temporary
** (Mix) Could not start application erlang_v8: could not find application file: erlang_v8.app
```

Did I miss something?

Here is my assumption. In short, we need to update elixir_v8 because of 404 error of `http://gyp.googlecode.com/svn/trunk`.

The current elixir_v8 fails building with erlang_v8 build failure log:

```
svn checkout --force http://gyp.googlecode.com/svn/trunk build/gyp \
    --revision 1831
svn: E170013: Unable to connect to a repository at URL 'http://gyp.googlecode.com/svn/trunk'
svn: E160013: '/svn/trunk' path not found
make[1]: *** [Makefile:443: dependencies] Error 1
```

So I updated elixir_v8 despite your `mix.lock`. Also I got another problem of building erlang_v8 so [I asked and had them fix](https://github.com/strange/erlang_v8/issues/4).

Now I got old elixir_v8 and new erlang_v8. The API of erlang_v8 that elixir_v8 expects is changed, so we got `** (ArgumentError) argument error`.

Let me check further.

Edit: They fixed it https://github.com/strange/erlang_v8/issues/3

--- a/lib/elixir_v8.ex
+++ b/lib/elixir_v8.ex
@@ -30,15 +30,17 @@ defmodule ElixirV8 do
   end
 
   def eval(pool_name, source) do
-    eval(pool_name, source, timeout)
+    eval(pool_name, source, timeout())
   end
 
   def eval(pool_name, source, timeout) do
-    :poolboy.transaction(pool_name, fn(worker) -> :erlang_v8_vm.eval(worker, source, timeout) end)
+    :poolboy.transaction(pool_name, fn(vm) ->
+      ElixirV8.VM.eval(vm, source, timeout)
+    end)
   end
 
   def eval_function(pool_name, source, args) do
-    eval_function(pool_name, source, args, timeout)
+    eval_function(pool_name, source, args, timeout())
   end
 
   def eval_function(pool_name, source, args, timeout) do
@@ -48,7 +50,7 @@ defmodule ElixirV8 do
         { :error, _ } -> []
       end
       js_function = "(function(){ var data = arguments[0]; " <> source <> "}).apply(null, " <> js_args <> ");"
-      :erlang_v8_vm.eval(worker, js_function, timeout)
+      ElixirV8.VM.eval(worker, js_function, timeout)
     end
     :poolboy.transaction(pool_name, f)
   end

--- a/lib/elixir_v8/supervisor.ex
+++ b/lib/elixir_v8/supervisor.ex
@@ -14,7 +14,7 @@ defmodule ElixirV8.Supervisor do
   def create_pool(pool_name, size, options) do
     args = [
       {:name, {:global, pool_name}},
-      {:worker_module, :erlang_v8_vm},
+      {:worker_module, ElixirV8.VM},
       {:size, size},
       {:max_overflow, 10}]
     pool_spec = :poolboy.child_spec(pool_name, args, options)
@@ -28,7 +28,10 @@ defmodule ElixirV8.Supervisor do
 
   def init([pools, global_or_local]) do
     spec_fun = fn({pool_name, pool_config}) ->
-      args = [{:name, {global_or_local, pool_name}}, {:worker_module, :erlang_v8_vm}] ++ pool_config
+      args = [
+        {:name, {global_or_local, pool_name}},
+        {:worker_module, ElixirV8.VM}
+      ] ++ pool_config
       :poolboy.child_spec(pool_name, args, pool_config)
     end
     pool_specs = Enum.map(pools, spec_fun)

--- a/lib/elixir_v8/vm.ex
+++ b/lib/elixir_v8/vm.ex
@@ -1,0 +1,23 @@
+defmodule ElixirV8.VM do
+  use GenServer
+
+  def start_link(opts) do
+    GenServer.start_link ElixirV8.VM, opts
+  end
+
+  def eval(vm, source, timeout) do
+    GenServer.call(vm, {:eval, source, timeout})
+  end
+
+  def handle_call({:eval, source, timeout}, _, {v8, context}) do
+    result = :erlang_v8.eval(v8, context, source, timeout)
+    {:reply, result, {v8, context}}
+  end
+
+  def init(opts) do
+    {:ok, v8} = :erlang_v8.start_vm(opts)
+    {:ok, context} = :erlang_v8.create_context(v8)
+    {:ok, {v8, context}}
+  end
+end
+

--- a/test/elixir_v8_test.exs
+++ b/test/elixir_v8_test.exs
@@ -1,7 +1,34 @@
 defmodule ElixirV8Test do
   use ExUnit.Case
+  doctest ElixirV8
 
-  test "the truth" do
-    assert 1 + 1 == 2
+  test "Simple execution" do
+    {:ok, _} = ElixirV8.create_pool(:main, 10)
+    {:ok, result} = ElixirV8.eval({:global, :main}, "2 * 4")
+    assert result == 8
+    ElixirV8.delete_pool(:main)
+  end
+
+  test "Context remaining" do
+    {:ok, _} = ElixirV8.create_pool(:main, 10)
+    {:ok, _} = ElixirV8.eval({:global, :main}, "function sum(a, b) { return a + b; }")
+    {:ok, result} = ElixirV8.eval({:global, :main}, "sum(3, 4)")
+    assert result == 7
+    ElixirV8.delete_pool(:main)
+  end
+
+  test "Function calling" do
+    {:ok, _} = ElixirV8.create_pool(:main, 10)
+    {:ok, result} = ElixirV8.eval_function({:global, :main}, "return arguments[0] * arguments[1]", [20, 5])
+    assert result == 100
+    ElixirV8.delete_pool(:main)
+  end
+
+  test "Loading JavaScript library" do
+    js_library = Path.join File.cwd!, "test/resource/library.js"
+    {:ok, _} = ElixirV8.create_pool(:main, 10, file: js_library)
+    {:ok, result2} = ElixirV8.eval({:global, :main}, "multiply(3, 4)")
+    assert result2 == 12
+    ElixirV8.delete_pool(:main)
   end
 end

--- a/test/resource/library.js
+++ b/test/resource/library.js
@@ -1,0 +1,3 @@
+function multiply(a, b) {
+  return a * b;
+}

